cli/azd/cmd/cobra_builder.go

```diff
@@ -58,28 +59,58 @@ func (cb *CobraBuilder) BuildCommand(descriptor *actions.ActionDescriptor) (*cob
 		}
 	}
 
-	// Configure action resolver for leaf commands
-	if !cmd.HasSubCommands() {
-		if err := cb.configureActionResolver(cmd, descriptor); err != nil {
-			return nil, err
-		}
+	// Configure action resolver for all commands
+	if err := cb.configureActionResolver(cmd, descriptor); err != nil {
+		return nil, err
 	}
 
 	return cmd, nil
 }
 
+func handleDocsFlag(
+	ctx context.Context, cmd *cobra.Command, container *ioc.NestedContainer) (bool, error) {
+	// Handle --docs flags for all commands
+	// Each command can use the description.options.DocumentationUrl to set what url to browse.
+	// By default, azd will use documentationHostName
+	if openDocs, _ := cmd.Flags().GetBool("docs"); openDocs {
+		var console input.Console
+		err := container.Resolve(&console)
+		if err != nil {
+			return false, err
+		}
+
+		parent := cmd.Parent()
+		commandPath := cmd.Use
+		for parent != nil {
+			commandPath = parent.Use + "-" + commandPath
+			parent = parent.Parent()
+		}
+
+		commandDocsUrl := documentationHostName + "/reference#" + commandPath
+		OpenWithDefaultBrowser(ctx, console, commandDocsUrl)
+		return true, nil
+	}
+	return false, nil
+}
```

Comment:
This is a great use case to be implemented as a middleware component vs having to change how a flag influences the behavior of all command behavior.

Reply:
AFAK, middleware can only work when there is an `action descriptor` to define what a command does.
If a command uses `RunE` directly, of has neither `RunE` or `action descriptor`, middleware can't be attached.

For example, if you run `AZD_DEBUG=true azd` or `AZD_DEBUG=true azd pipeline`, the debug middleware is never hit.

In case of `--docs` flag, we want to support it for runnable and non-runnable commands, like `azd --docs` and `azd pipeline --docs`.  
I considered creating an Action descriptor for all non-runnable commands, but it doesn't feel like is worth it, as they only need to either print cmd.Help or honor the `--docs` flags if present.

The other reason why I think a middleware should not be used in this case is because a middleware is invoked after initializing the container and the options for the command. Because of that, running something like `azd provision --docs` would return an error if `azd` is run from a folder where there is not an azd-environment (azure.yaml), and similar with any other commands which has a dependency on either env/login/etc.  We would need to figure a way to call the middleware earlier.   The `--docs` command is more likely to the `--help` command, which is an early termination to any azd-core initialization.

Reply:
My slight preference is to avoid `cmd.Runnable()` returning true for commands with any "docs only behavior", which perhaps pushes us to a slightly more cobra-native implementation [here](https://github.com/Azure/azure-dev/pull/2463/files#r1245476835).